### PR TITLE
add separate unit tests for cluster status computation

### PR DESCRIPTION
**Pull request by anvddriesch**
Towards: https://github.com/giantswarm/giantswarm/issues/15468
Here we add small unit tests to check if the cluster status computations give the desired result.
For starters it includes the basic scenarios which are considered in the computation. Some are not covered yet and would fail (because the computation is actually broken) - let's add them as we fix things.

## Checklist

- [ ] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: service/controller/resource/statuscondition/create_test.go [hunk 1]
```diff
@@ -317,3 +317,370 @@ func newCommonClusterObjectFunc(provider string) func() infrastructurev1alpha2.C
...
+			name: "case 2",
+
+			status: infrastructurev1alpha2.CommonClusterStatus{
+				Conditions: []infrastructurev1alpha2.CommonClusterStatusCondition{
+					unittest.GetCreatingCondition(90),
+				},
+				Versions: []infrastructurev1alpha2.CommonClusterStatusVersion{},
+			},
+			expectedResult: false,
+		},
```
Comment (xh3b4sd) [inline service/controller/resource/statuscondition/create_test.go:361]: I would add more cases for versions and conditions being present. And then also multiples of that. Too much coverage is better than too little. Especially in projects with multiple and potentially changing contributors over time.

File: service/controller/resource/statuscondition/create_test.go [hunk 2]
```diff
@@ -317,3 +317,370 @@ func newCommonClusterObjectFunc(provider string) func() infrastructurev1alpha2.C
...
+					unittest.GetCreatingCondition(90),
+				},
+				Versions: []infrastructurev1alpha2.CommonClusterStatusVersion{
+					unittest.GetVersion(60, "8.7.6"),
+				},
+			},
+			desiredVersion: "8.7.6",
+
+			expectedResult: false,
+		},
```
Comment (xh3b4sd) [inline service/controller/resource/statuscondition/create_test.go:509]: I would add a case for desired version existing already while the updating condition is also already in place. The result should then be false as well.

File: service/controller/resource/statuscondition/create_test.go [hunk 3]
```diff
@@ -317,3 +317,370 @@ func newCommonClusterObjectFunc(provider string) func() infrastructurev1alpha2.C
...
+					unittest.GetCreatingCondition(90),
+				},
+			},
+			nodesReady: false,
+
+			expectedResult: false,
+		},
+		// the cluster is already updated
+		{
+			name: "case 1",
```
Comment (xh3b4sd) [inline service/controller/resource/statuscondition/create_test.go:563]:
```suggestion
			name: "case 2",
```

File: service/controller/resource/statuscondition/create_test.go [hunk 4]
```diff
@@ -317,3 +317,370 @@ func newCommonClusterObjectFunc(provider string) func() infrastructurev1alpha2.C
...
+				Versions: []infrastructurev1alpha2.CommonClusterStatusVersion{},
+			},
+			nodesReady:     true,
+			desiredVersion: "8.7.5",
+
+			expectedResult: false,
+		},
+		// the nodes are not ready yet
+		{
+			name: "case 0",
```
Comment (xh3b4sd) [inline service/controller/resource/statuscondition/create_test.go:637]:
```suggestion
			name: "case 2",
```

File: service/controller/resource/statuscondition/create_test.go [hunk 5]
```diff
@@ -317,3 +317,370 @@ func newCommonClusterObjectFunc(provider string) func() infrastructurev1alpha2.C
...
+				},
+			},
+			nodesReady:     false,
+			desiredVersion: "8.7.6",
+
+			expectedResult: false,
+		},
+		// the version is already set
+		{
+			name: "case 0",
```
Comment (xh3b4sd) [inline service/controller/resource/statuscondition/create_test.go:657]:
```suggestion
			name: "case 3",
```

**Comment by anvddriesch** [comment]
I added some more cases as suggested. There are also a couple more I have in mind which would currently fail so I'll add them when fixing the status computation.

**Review by xh3b4sd — APPROVED** [review]

**Pull request merged**